src/widget/tree_structure.rs

```diff
@@ -0,0 +1,155 @@
+use crate::id::Id;
+use std::collections::HashMap;
+
+/// The pure structure (parent/children relations via ids) of the widget tree.
+#[derive(Debug, Default, Clone)]
+pub struct TreeStructure {
+    parent: HashMap<Id, Option<Id>>,
+    // TODO this is taken from the druid lasagna branch, is the Option here intentional, i.e. to track all parentless ids with None?
+    children: HashMap<Option<Id>, Vec<Id>>,
+}
+
+impl TreeStructure {
+    pub fn parent(&self, id: Id) -> Option<Option<Id>> {
+        self.parent.get(&id).copied()
+    }
+
+    pub fn children(&self, id: Option<Id>) -> Option<&[Id]> {
+        self.children.get(&id).map(Vec::as_slice)
+    }
+
+    pub fn is_descendant_of(&self, mut id: Id, ancestor: Id) -> bool {
+        while let Some(parent) = self.parent(id).flatten() {
+            if parent == ancestor {
+                return true;
+            }
+            id = parent;
+        }
+        false
+    }
+
+    // TODO: Should any of the following methods return a Result or panic?
+    pub(crate) fn append_child(&mut self, parent_id: Id, id: Id) {
```

Comment:
Panic for now.

Code that fails silently can stay under the radar for years. Code that panics gets bug reports fast, and panics are easier to find in unit tests and fuzz tests. Returning Result doesn't make sense, because any id-related error here will be because of an faulty implementation in Xilem, not user code or bad IO.

(Maybe mention the panics in doc comments, though.)